File: lib/generate/tables/maxp.js
'use strict';

// See documentation here: http://www.microsoft.com/typography/otspec/maxp.htm

var _ = require('underscore');
var ByteBuffer = require('../byte_buffer.js');

// Find max points in glyph TTF contours.
function getMaxPoints(glyphs) {
    return 0;
//    return _.max(_.map(glyphs, function (glyph) {
//        return glyph.points.length;
//    }));
}

function getMaxContours(glyphs) {
    return 0;
//    return _.max(_.map(glyphs, function (glyph) {
//        return glyph.ttfContours.length;
//    }));
}

function createMaxpTable(font) {
    var glyphs = font._glyphs;
    var buf = ByteBuffer.prototype.create(32);
    var maxp = font.tables.maxp;

    buf.writeInt32(maxp.version); // version
    buf.writeUint16(glyphs.length); // numGlyphs
    buf.writeUint16(maxp.maxPoints);// getMaxPoints(glyphs)); // maxPoints
    buf.writeUint16(maxp.maxContours); //getMaxContours(glyphs)); // maxContours
    buf.writeUint16(maxp.maxCompositePoints); // maxCompositePoints
    buf.writeUint16(maxp.maxCompositeContours); // maxCompositeContours
    buf.writeUint16(maxp.maxZones); // maxZones
    buf.writeUint16(maxp.maxTwilightPoints); // maxTwilightPoints
    // It is unclear how to calculate maxStorage, maxFunctionDefs and maxInstructionDefs.
    // These are magic constants now, with values exceeding values from FontForge
    buf.writeUint16(maxp.maxStorage); // maxStorage
    buf.writeUint16(maxp.maxFunctionDefs); // maxFunctionDefs
    buf.writeUint16(maxp.maxInstructionDefs); // maxInstructionDefs
    buf.writeUint16(maxp.maxStackElements); // maxStackElements
    buf.writeUint16(maxp.maxSizeOfInstructions); // maxSizeOfInstructions
    buf.writeUint16(maxp.maxComponentElements); // maxComponentElements
    buf.writeUint16(maxp.maxComponentDepth); // maxComponentDepth

    return buf;
}

module.exports = createMaxpTable;

File: src/node/tables/maxp.cc
#include <nan.h>
#include <ft2build.h>
#include "../freetype.h"
#include FT_TRUETYPE_TABLES_H

NAN_METHOD(FreeType::GetMaxProfileTable) {
    NanScope();
    FreeType* ff = ObjectWrap::Unwrap<FreeType>(args.This());
    TT_MaxProfile* maxp = (TT_MaxProfile*) FT_Get_Sfnt_Table(ff->face, ft_sfnt_maxp);
    if(maxp == NULL) {
        NanReturnUndefined();
    }
    Local<Object> obj = NanNew<Object>();

//typedef struct  TT_MaxProfile_
//  {
//    FT_Fixed   version;
//    FT_UShort  numGlyphs;
//    FT_UShort  maxPoints;
//    FT_UShort  maxContours;
//    FT_UShort  maxCompositePoints;
//    FT_UShort  maxCompositeContours;
//    FT_UShort  maxZones;
//    FT_UShort  maxTwilightPoints;
//    FT_UShort  maxStorage;
//    FT_UShort  maxFunctionDefs;
//    FT_UShort  maxInstructionDefs;
//    FT_UShort  maxStackElements;
//    FT_UShort  maxSizeOfInstructions;
//    FT_UShort  maxComponentElements;
//    FT_UShort  maxComponentDepth;
//
//  } TT_MaxProfile;

    OBJ_SET_INTEGER("version", maxp->version);
    OBJ_SET_INTEGER("numGlyphs", maxp->numGlyphs);
    OBJ_SET_INTEGER("maxPoints", maxp->maxPoints);
    OBJ_SET_INTEGER("maxContours", maxp->maxContours);
    OBJ_SET_INTEGER("maxCompositePoints", maxp->maxCompositePoints);
    OBJ_SET_INTEGER("maxCompositeContours", maxp->maxCompositeContours);
    OBJ_SET_INTEGER("maxZones", maxp->maxZones);
    OBJ_SET_INTEGER("maxTwilightPoints", maxp->maxTwilightPoints);
    OBJ_SET_INTEGER("maxStorage", maxp->maxStorage);
    OBJ_SET_INTEGER("maxFunctionDefs", maxp->maxFunctionDefs);
    OBJ_SET_INTEGER("maxInstructionDefs", maxp->maxInstructionDefs);
    OBJ_SET_INTEGER("maxStackElements", maxp->maxStackElements);
    OBJ_SET_INTEGER("maxSizeOfInstructions", maxp->maxSizeOfInstructions);
    OBJ_SET_INTEGER("maxComponentElements", maxp->maxComponentElements);
    OBJ_SET_INTEGER("maxComponentDepth", maxp->maxComponentDepth);

    NanReturnValue(obj);
}